src/client.ts

```diff
@@ -194,21 +220,25 @@ export class Client {
   public async close(
     params: { closingCode: number; closingReason: string } = { closingCode: 0, closingReason: "" }
   ): Promise<void> {
-    this.logger.info(`Closing client...`)
-    if (this.publisherCounts()) {
-      this.logger.info(`Stopping all publishers...`)
-      await this.closeAllPublishers()
-    }
-    if (this.consumerCounts()) {
-      this.logger.info(`Stopping all consumers...`)
-      await this.closeAllConsumers()
+    this.logger.info(`${this.id} Closing client...`)
+    const refs = this.decrRefCount()
+    if (refs <= 0) {
+      if (this.publisherCounts()) {
+        this.logger.info(`Stopping all producers...`)
+        await this.closeAllPublishers()
+      }
+      if (this.consumerCounts()) {
+        this.logger.info(`Stopping all consumers...`)
+        await this.closeAllConsumers()
+      }
+      this.logger.info(`Stopping heartbeat...`)
+      this.heartbeat.stop()
+      this.logger.debug(`Close...`)
+      const closeResponse = await this.sendAndWait<CloseResponse>(new CloseRequest(params))
+      this.logger.debug(`Close response: ${closeResponse.ok} - '${inspect(closeResponse)}'`)
+      Client.removeCachedClient(this)
```

Comment:
having a static method that take the instance as a parameter seems odd but I think that is a smell of client having too much responsibilities